clippy_lints/src/try_err.rs

```diff
@@ -0,0 +1,120 @@
+use crate::utils::{match_qpath, paths, snippet, span_lint_and_then};
+use if_chain::if_chain;
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::ty::Ty;
+use rustc::{declare_lint_pass, declare_tool_lint};
+use rustc_errors::Applicability;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for usages of `Err(x)?`.
+    ///
+    /// **Why is this bad?** The `?` operator is designed to allow calls that
+    /// can fail to be easily chained. For example, `foo()?.bar()` or
+    /// `foo(bar()?)`. Because `Err(x)?` can't be used that way (it will
+    /// always return), it is more clear to write `return Err(x)`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// // Bad
+    /// fn foo(fail: bool) -> Result<i32, String> {
+    ///     if fail {
+    ///       Err("failed")?;
+    ///     }
+    ///     Ok(0)
+    /// }
+    ///
+    /// // Good
+    /// fn foo(fail: bool) -> Result<i32, String> {
+    ///     if fail {
+    ///       return Err("failed".into());
+    ///     }
+    ///     Ok(0)
+    /// }
+    /// ```
+    pub TRY_ERR,
+    style,
+    "return errors explicitly rather than hiding them behind a `?`"
+}
+
+declare_lint_pass!(TryErr => [TRY_ERR]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TryErr {
```

Comment:
I think this lint is easier implemented as an `EarlyLintPass`, when the `?` is not desugared.

```rust
if_chain! {
    if let ExprKind::Try(err_expr) = &expr.node;
    if let ExprKind::Call(func, _) = &err_expr.node;
    if let ExprKind::Path(_, path) = &func.node;
    if match_path(path, ERR_PATH); // or something like this
    then {
        // ...
    }
}
```

Reply:
Thanks, I'll try it. One thing I'm not sure of though: right now, I'm conditionally suggesting `.into()` only if it's necessary. To do that, I'm inspecting the resulting type for a couple of expressions. Is there a way to do that in the early pass? Or, is there some other way to detect that `.into()` will be required?

Reply:
Oh, I didn't consider that. No type checking is done after Early Lints are run, therefore you can't compare the types.

So using a `LateLintPass` is totally fine here.